docs/_docs/02_features/provider-cache.md

```diff
@@ -0,0 +1,132 @@
+TERRAGRUNT_PROVIDER_CACHE=1 \
+TERRAGRUNT_PROVIDER_CACHE_HOST=192.168.0.100 \
+TERRAGRUNT_PROVIDER_CACHE_PORT=5758 \
+TERRAGRUNT_PROVIDER_CACHE_TOKEN=my-secret \
+terragrunt apply
+```
+
+### How Terragrunt Provider Caching works
+
+* Run Terragurn Provider Cache server on localhost.
+* Configure Terraform instances to use the cache server as a remote registry:
+  * Create local CLI config file `.terragrunt-cache/.terraformrc` that concatenates the user configuration from the Terraform [CLI config file](https://developer.hashicorp.com/terraform/cli/config/config-file) with additional sections:
+     * [provider-installation](https://developer.hashicorp.com/terraform/cli/config/config-file#provider-installation) forces Terraform to look for for the required providers in the cache directory and create symbolic links to them, if not found, then request them from the remote registry.
+     * [host](https://github.com/hashicorp/terraform/issues/28309) forces Terraform to [forward](#How forwarding Terraform requests through the cache server works) all provider requests through the cache server.
+  * Set environment variables:
+     * [TF_PLUGIN_CACHE_MAY_BREAK_DEPENDENCY_LOCK_FILE](https://developer.hashicorp.com/terraform/cli/config/config-file#allowing-the-provider-plugin-cache-to-break-the-dependency-lock-file) allows to generate `.terraform.lock.hcl` files based only on provider hashes from the cache directory.
+     * [TF_CLI_CONFIG_FILE](https://developer.hashicorp.com/terraform/cli/config/environment-variables#tf_plugin_cache_dir) sets to use just created local CLI config `.terragrunt-cache/.terraformrc`
+     * [TF_TOKEN_*](https://developer.hashicorp.com/terraform/cli/config/config-file#environment-variable-credentials) sets per-remote-registry tokens for authentication to cache server.
+* Call `terraform providers lock -platform=cache_provider`. When the cache server receives this request, it returns HTTP status _429 Locked_ and starts caching the provider (for any other requests, cache server acts as a proxy):
+  * Create the [lock file](https://en.wikipedia.org/wiki/File_locking) to prevent multiple cache servers from overwriting the same provider cache.
+  * Download the provider from the remote registry, unpack and store into the cache directory or [create a symlink](#Why Cache Server creates symlinks to providers from the user plugins directory) if the required provider exists in the user plugins directory.
+  * Remove the [lock file](https://en.wikipedia.org/wiki/File_locking).
+* Receive the response _429 Locked_ from the cache server, wait until all providers are cached.
+* If the [`terragrunt-provider-cache-complete-lock`](https://terragrunt.gruntwork.io/docs/reference/cli-options/#terragrunt-provider-cache-complete-lock) flag is set, call `terraform providers lock` to generate a complete `.terraform.lock.hcl` file.
+* Run `terragrunt init`, which in turn finds providers in the cache directory and creates symlinks to them.
```

Comment:
```suggestion
* Start a server on localhost. This is the _Terragrunt Provider Cache server_.
* Configure Terraform instances to use the cache server as a remote registry:
  * Create local CLI config file `.terragrunt-cache/.terraformrc` that concatenates the user configuration from the Terraform [CLI config file](https://developer.hashicorp.com/terraform/cli/config/config-file) with additional sections:
     * [provider-installation](https://developer.hashicorp.com/terraform/cli/config/config-file#provider-installation) forces Terraform to look for for the required providers in the cache directory and create symbolic links to them, if not found, then request them from the remote registry.
     * [host](https://github.com/hashicorp/terraform/issues/28309) forces Terraform to [forward](#how-forwarding-terraform-requests-through-the-cache-server-works) all provider requests through the cache server.
  * Set environment variables:
     * [TF_PLUGIN_CACHE_MAY_BREAK_DEPENDENCY_LOCK_FILE](https://developer.hashicorp.com/terraform/cli/config/config-file#allowing-the-provider-plugin-cache-to-break-the-dependency-lock-file) allows to generate `.terraform.lock.hcl` files based only on provider hashes from the cache directory.
     * [TF_CLI_CONFIG_FILE](https://developer.hashicorp.com/terraform/cli/config/environment-variables#tf_plugin_cache_dir) sets to use just created local CLI config `.terragrunt-cache/.terraformrc`
     * [TF_TOKEN_*](https://developer.hashicorp.com/terraform/cli/config/config-file#environment-variable-credentials) sets per-remote-registry tokens for authentication to cache server.
* Any time Terragrunt is going to run `init`:
    * Call `terraform providers lock -platform=cache_provider`. This gets Terraform to request all the providers it needs from the Terragrunt Provider Cache server. 
    * The Terragrunt Provider Cache server will download the provider from the remote registry, unpack and store it into the cache directory or [create a symlink](#Why Cache Server creates symlinks to providers from the user plugins directory) if the required provider exists in the user plugins directory. Note that the Terragrunt Provider Cache server will ensure that each unique provider is only ever downloaded and stored on disk once, handling concurrency (from multiple Terraform and Terragrunt instances) correctly. 
    * The Terragrunt Provider Cache server returns the HTTP status _429 Locked_ to Terraform. This is because we do _not_ want Terraform to actually download any providers as a result of calling `terraform providers lock` (as that command does _not_ use any caching); we only use that command to get the list of required providers.
    * At this point, all providers are downloaded and cached, so finally, we run `terragrunt init`, which will find all the providers it needs in the cache, and it'll create symlinks to them nearly instantly, with no additional downloading.
    * Note that if a Terraform module doesn't have a lock file, Terraform does _not_ use the cache, so it would end up downloading all the providers from scratch. To work around this, by default, we use the `plugin_cache_may_break_dependency_lock_file` feature, which, for modules without lock files, will allow Terraform to automatically generate a partial lock file if it finds the providers it needs in the cache, with no additional downloading. This ensures that no additional downloads are necessary, but at the cost of partial lock files. If you wish to disable this feature, set the [`terragrunt-provider-cache-complete-lock`](https://terragrunt.gruntwork.io/docs/reference/cli-options/#terragrunt-provider-cache-complete-lock) flag, and for modules without a lock file, Terragrunt will call `terraform providers lock`, which will give you a complete lock file, at the cost of more downloading and bandwidth.
```

Reply:
Thanks.